run/fvde2john.py

```diff
@@ -0,0 +1,220 @@
+
+            return volume_unlock_record
+
+    return None
+
+def get_all_partitions_of_type(partition_dict, part_type):
+    return [partition_dict[p]['start'] for p in partition_dict if partition_dict[p]['partition_type'] == guid_to_hex(part_type)]
+
+def load_plist_dict(pt):
+    # resultant pt has one extra malformed line in the xml, so we remove this.
+    plist_str = b''.join(pt.split(b'\n')[:-1]).decode()
+    d = plistlib.loads(plist_str)
+    return d
+
+# Recursive traversal - Recovery HD partition does not contain a lot of files, and so this approach is fine
+def traverse_filesystem(fs_object, target_file, path='/'):
+    for entry in fs_object.open_dir(path):
+        try:
+            if entry.info.name.name in [b'.', b'..']:
+                continue
+
+            file_path = os.path.join(path, entry.info.name.name.decode('utf-8'))
+
+            if entry.info.meta and entry.info.meta.type == pytsk3.TSK_FS_META_TYPE_REG:
+                if entry.info.name.name == target_file.encode():
+                    print(f"[+] Found file: {file_path}")
+                    file_data = recover_file(fs_object, file_path)
+
+                    # this returns to previous call i.e. the dir layer
+                    return file_data
+
+            # Traverse lower layer if entry is a dir
+            elif entry.info.meta and entry.info.meta.type == pytsk3.TSK_FS_META_TYPE_DIR:
+                file_data = traverse_filesystem(fs_object, target_file, file_path)
+
+                if file_data:
+                    return file_data
+
+        except Exception as e:
+            print(f"[!] Error reading entry: {e}")
+
+def recover_file(fs_object, file_path):
+    try:
+        file_obj = fs_object.open(file_path)
+        size = file_obj.info.meta.size
+        offset = 0
+        data = file_obj.read_random(offset, size)
+
+        return data
+
+    except Exception as e:
+        print(f"[!] Error recovering file: {e}")
+
+def get_EncryptedRoot_plist_wipekey(image_file, start_pos):
+    img = pytsk3.Img_Info(image_file)
+    fs = pytsk3.FS_Info(img, offset=start_pos)
+    target_file = 'EncryptedRoot.plist.wipekey'
+    EncryptedRoot_data = traverse_filesystem(fs, target_file)
+
+    return EncryptedRoot_data
+
+def main():
+
+    p = argparse.ArgumentParser()
+    p.add_argument('image_file')
+    args = p.parse_args()
+    image_file = args.image_file
+
+    with open(image_file, 'rb') as fp:
+        partition_dict = parse_partition_table(fp)
+
+        core_storage_volumes = get_all_partitions_of_type(partition_dict, HEX_CORE_STORAGE_TYPE_GUID)
+        boot_volumes = get_all_partitions_of_type(partition_dict, HEX_APPLE_BOOT_STORAGE_TYPE_GUID)
+
+        # Unlikely to have more than one boot volume, but loop anyway
+        for boot_start_pos in boot_volumes:
+            EncryptedRoot_data = get_EncryptedRoot_plist_wipekey(image_file, boot_start_pos)
+
+            if core_storage_volumes == []:
```

Comment:
Since this condition is already clear before entering the for loop, I suggest moving it before the loop.
Besides, the `exit()` should be a `sys.exit(1)`.